### PR TITLE
Performance Fixes

### PR DESCRIPTION
**Pull request by joewalsh**
- Move face detection to the GPU and reduce accuracy
- Eliminate main thread file writes for article loads, cache frequently used proxied files in memory
- Fix bug where location monitoring was on constantly from app launch even if a location section wasn't viewed

### CODE REVIEW

**Comment by joewalsh** [comment]
[5 Obama Article Loads.trace.zip](https://github.com/wikimedia/wikipedia-ios/files/361537/5.Obama.Article.Loads.trace.zip)

^ two runs of searching for the obama article then loading it five times. first run is develop branch, second run is this branch. Less CPU usage overall, not sure if it's really user-noticeable in terms of article load speed. Over the long run, less CPU would mean less battery consumption.

In addition, eliminating the unnecessary fine grained location tracking should have a much larger impact in terms of reducing battery consumption.

**Comment by joewalsh** [comment]
^ image is centered properly but test fails. will update

File: Wikipedia/Code/CIDetector+WMFFaceDetection.m [hunk 1]
```diff
@@ -1,27 +1,21 @@
...
-+ (instancetype)wmf_sharedBackgroundFaceDetector {
++ (instancetype)wmf_sharedFaceDetector {
     static CIDetector* defaultFaceDetector;
     static dispatch_once_t onceToken;
     dispatch_once(&onceToken, ^{
         defaultFaceDetector = [CIDetector detectorOfType:CIDetectorTypeFace
-                                                 context:[CIContext wmf_sharedBackgroundContext]
-                                                 options:nil];
+                                                 context:[CIContext wmf_sharedContext]
+                                                 options:@{CIDetectorAccuracy: CIDetectorAccuracyLow}];
```
Comment (coreyfloyd) [inline Wikipedia/Code/CIDetector+WMFFaceDetection.m]: Did we verify this being OK for our images… any major differences in face detection?

Comment (joewalsh) [inline Wikipedia/Code/CIDetector+WMFFaceDetection.m]: No major difference that I noticed. The only test that changed slightly was the Obama one - it's still correct but a few pixels offset.

Comment (coreyfloyd) [inline Wikipedia/Code/CIDetector+WMFFaceDetection.m]: 👍

**Comment by coreyfloyd** [comment]
@joewalsh for some reason that trace is crashing in my version of instruments… can you give the high level details of the performance improvement?

**Comment by joewalsh** [comment]
screen shot 2016-07-14 at 11 37 05 am

screen shot 2016-07-14 at 11 37 15 am

@coreyfloyd ^ less CPU usage overall. no noticeable article load speedup.

**Comment by coreyfloyd** [comment]
👍

Also thats a lot of emoji in that device name.

File: Wikipedia/Code/WKWebView+LoadAssetsHtml.m [hunk 2]
```diff
@@ -47,19 +53,12 @@ - (void)loadHTML:(NSString*)string baseURL:(NSURL *)baseURL withAssetsFile:(NSSt
     // index.html and preview.html have four "%@" subsitition markers. Replace both of these with actual content.
     NSString* templateAndContent = [NSString stringWithFormat:fileContents, fontString, baseURL.absoluteString, @(topPadding), string];
 
-    // Get temp file name. For a fileName of "index.html" the temp file name would be "index.temp.html"
-    NSString* tempFileName = [[[fileName stringByDeletingPathExtension] stringByAppendingPathExtension:@"temp"] stringByAppendingPathExtension:[fileName pathExtension]];
-
-    // Get path to tempFileName
-    NSString* tempFilePath = [proxyServer localFilePathForRelativeFilePath:tempFileName];
+    
+    [proxyServer setResponseData:[templateAndContent dataUsingEncoding:NSUTF8StringEncoding] withContentType:@"text/html; charset=utf-8" forPath:fileName];
```
Comment (coreyfloyd) [inline Wikipedia/Code/WKWebView+LoadAssetsHtml.m]: Oh so by setting this response data we no longer have to write a temp file?

File: Wikipedia/Code/WKWebView+LoadAssetsHtml.m [hunk 3]
```diff
@@ -16,11 +18,12 @@ - (void)loadHTMLFromAssetsFile:(NSString*)fileName scrolledToFragment:(NSString*
     
     fragment = fragment ? fragment : @"top";
     NSURL* requestURL = [[WMFProxyServer sharedProxyServer] proxyURLForRelativeFilePath:fileName fragment:fragment];
-    NSURLRequest* request = [NSURLRequest requestWithURL:requestURL];
+    NSURLRequest* request = [NSURLRequest requestWithURL:requestURL cachePolicy:NSURLRequestReloadIgnoringCacheData timeoutInterval:WKWebViewLoadAssetsHTMLRequestTimeout];
```
Comment (coreyfloyd) [inline Wikipedia/Code/WKWebView+LoadAssetsHtml.m]: Are we passing `NSURLRequestReloadIgnoringCacheData` now so it always gets whatever response you set in:
 
` [proxyServer setResponseData:[templateAndContent dataUsingEncoding:NSUTF8StringEncoding] withContentType:@"text/html; charset=utf-8" forPath:fileName];`

File: Wikipedia/Code/WKWebView+LoadAssetsHtml.m (hunk 2 above)
Comment (joewalsh) [inline Wikipedia/Code/WKWebView+LoadAssetsHtml.m]: correct

File: Wikipedia/Code/WKWebView+LoadAssetsHtml.m (hunk 3 above)
Comment (joewalsh) [inline Wikipedia/Code/WKWebView+LoadAssetsHtml.m]: yeah- it should have had `NSURLRequestReloadIgnoringCacheData` even in the file case

**Comment by coreyfloyd** [comment]
Just merging it already passed before updating with develop…

**Comment by montehurd** [comment]
For the record, this PR was great!

**Pull request merged**